src/session.c

```diff
@@ -443,10 +492,19 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
             pinlen = cb_pin_len;
 
             cache = p11prov_ctx_cache_pins(session->provctx);
-
         } else {
-            ret = CKR_CANCEL;
-            goto done;
+            /* We are asking the user off-band for the user consent -- from store
+             * we will always receive non-null (but unusable) callback */
+            ret = p11prov_session_get_pin(slot, (char *)cb_pin, &cb_pin_len);
+            if (ret != RET_OSSL_OK) {
+                ret = CKR_GENERAL_ERROR;
+                goto done;
+            }
+
+            pin = (CK_UTF8CHAR_PTR)cb_pin;
+            pinlen = strlen(cb_pin);
+
+            cache = p11prov_ctx_cache_pins(session->provctx);
```

Comment:
maybe move this after the if/else block so we do it only once ?

Reply:
Done.

Reply:
Ok, this wont work as it tries to cache the cached pin, which goes into user-after-free. Indenting this one more tab will make it more ugly, but works. We can reiterate later though.